Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -14,38 +15,179 @@ ComponentInfo::ComponentInfo()
+  const auto range = (*m_detectorRanges)[rangesIndex];
   // Extract as a block
   return std::vector<size_t>(
       m_assemblySortedDetectorIndices->begin() + range.first,
       m_assemblySortedDetectorIndices->begin() + range.second);
 }
 
+std::vector<size_t>
+ComponentInfo::componentIndices(const size_t componentIndex) const {
+  if (isDetector(componentIndex)) {
+    /* This is a single detector. Just return the corresponding index.
+     * detectorIndex == componentIndex. Never any sub-components for detectors.
+     */
+    return std::vector<size_t>{componentIndex};
+  }
+  // Calculate index into our ranges (non-detector) component items.
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  const auto range = (*m_componentRanges)[rangesIndex];
+
+  // Extract as a block
+  return std::vector<size_t>(
+      m_assemblySortedComponentIndices->begin() + range.first,
+      m_assemblySortedComponentIndices->begin() + range.second);
+}
+
 size_t ComponentInfo::size() const { return m_size; }
 
 bool ComponentInfo::operator==(const ComponentInfo &other) const {
-  return m_size == other.m_size && m_ranges == other.m_ranges &&
+  return m_size == other.m_size && m_detectorRanges == other.m_detectorRanges &&
          *m_assemblySortedDetectorIndices ==
              *other.m_assemblySortedDetectorIndices;
 }
 bool ComponentInfo::operator!=(const ComponentInfo &other) const {
   return !(this->operator==(other));
 }
+
+Eigen::Vector3d ComponentInfo::position(const size_t componentIndex) const {
+  if (isDetector(componentIndex)) {
+    return m_detectorInfo->position(componentIndex);
+  }
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  return (*m_positions)[rangesIndex];
+}
+
+Eigen::Quaterniond ComponentInfo::rotation(const size_t componentIndex) const {
+  if (isDetector(componentIndex)) {
+    return m_detectorInfo->rotation(componentIndex);
+  }
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  return (*m_rotations)[rangesIndex];
+}
+
+/**
+ * Sets the rotation for a component described by target component index
+ *
+ * This will propagate and apply the derived position offsets to all known
+ *sub-components
+ *
+ * @param componentIndex : Component index to update at
+ * @param newPosition : Absolute position to set
+ */
+void ComponentInfo::setPosition(const size_t componentIndex,
+                                const Eigen::Vector3d &newPosition) {
+
+  const Eigen::Vector3d offset = newPosition - position(componentIndex);
+  const auto indices = this->componentIndices(
+      componentIndex); // Includes requested component index
+  for (auto &index : indices) {
+    if (isDetector(index)) {
+      m_detectorInfo->setPosition(index,
+                                  m_detectorInfo->position(index) + offset);
+    } else {
+      m_positions.access()[compOffsetIndex(index)] += offset;
+      // TODO. This could change L1/L2. Is that being detected?
```

Comment:
Now that we are storing the position of source and sample in `ComponentInfo` the positions access is fast and thus the L1/L2 caching mechanism should be removed from `DetectorInfo`.